presto-main/src/main/java/com/facebook/presto/operator/aggregation/state/BlockPositionValueStateSerializer.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.state;
+
+import com.facebook.presto.spi.block.AbstractRowBlock;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.ColumnarRow;
+import com.facebook.presto.spi.function.AccumulatorStateSerializer;
+import com.facebook.presto.spi.type.RowType;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.spi.block.ColumnarRow.toColumnarRow;
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.google.common.base.Preconditions.checkArgument;
+
+public abstract class BlockPositionValueStateSerializer<T extends BlockPositionValueState>
+        implements AccumulatorStateSerializer<T>
+{
+    protected final Type firstType;
+
+    protected final Type secondType;
+
+    abstract void readFirstField(Block block, int index, T state);
+
+    abstract void writeFirstField(BlockBuilder out, T state);
+
+    abstract void writeSecondField(BlockBuilder out, Block block, int position);
```

Comment:
The reason why you need this as an abstract method (and thus the abstract classes `BlockValueStateSerializer` and `SliceValueStateSerializer` is because you want to call `writeObject` / `writeSlice`.

However, you can do this
```
        secondType.appendTo(block, position, out);
```

Do this, inline `writeSecondField`. And remove `BlockValueStateSerializer` and `SliceValueStateSerializer`.